File: internal/radio/handler/radio.go
package handler

import (
	"net/http"

	"github.com/arung-agamani/denpa-radio/internal/apiresponse"
	"github.com/arung-agamani/denpa-radio/internal/radio/service"
	"github.com/gin-gonic/gin"
)

// RadioHandlers holds the gin route handlers for station status, scheduler,
// and skip endpoints.
type RadioHandlers struct {
	svc *service.RadioService
}

func NewRadioHandlers(svc *service.RadioService) *RadioHandlers {
	return &RadioHandlers{svc: svc}
}

// Health handles GET /health
func (h *RadioHandlers) Health(c *gin.Context) {
	c.JSON(http.StatusOK, gin.H{"status": "ok"})
}

// Status handles GET /api/status
func (h *RadioHandlers) Status(c *gin.Context) {
	snap := h.svc.Status()
	var currentTrackInfo interface{}
	if snap.CurrentTrackRaw != nil {
		currentTrackInfo = sanitiseTrack(snap.CurrentTrackRaw)
	}
	apiresponse.OK(c, gin.H{
		"station_name":       snap.StationName,
		"current_track":      snap.CurrentTrack,
		"current_track_info": currentTrackInfo,
		"total_tracks":       snap.TotalTracks,
		"library_tracks":     snap.LibraryTracks,
		"active_clients":     snap.ActiveClients,
		"max_clients":        snap.MaxClients,
		"active_tag":         snap.ActiveTag,
		"active_playlist":    snap.ActivePlaylist,
		"active_playlist_id": snap.ActivePlaylistID,
		"scheduler_running":  snap.SchedulerRunning,
		"playlist_summary":   snap.PlaylistSummary,
		"timezone":           snap.Timezone,
		"server_time":        snap.ServerTime,
	})
}

// SchedulerStatus handles GET /api/scheduler/status
func (h *RadioHandlers) SchedulerStatus(c *gin.Context) {
	snap := h.svc.SchedulerStatus()
	apiresponse.OK(c, gin.H{
		"running":        snap.Running,
		"last_tag":       snap.LastTag,
		"time_tags":      snap.TimeTags,
		"current_tag":    snap.CurrentTag,
		"summary":        snap.Summary,
		"library_tracks": snap.LibraryTracks,
		"timezone":       snap.Timezone,
		"server_time":    snap.ServerTime,
	})
}

// GetQueue handles GET /api/queue  (public)
func (h *RadioHandlers) GetQueue(c *gin.Context) {
	tracks := h.svc.GetQueue(0)
	apiresponse.OK(c, gin.H{
		"tracks": sanitiseTracks(tracks),
	})
}

// SkipNext handles POST /api/skip/next  (protected)
func (h *RadioHandlers) SkipNext(c *gin.Context) {
	h.svc.SkipNext()
	apiresponse.OK(c, gin.H{})
}

// SkipPrev handles POST /api/skip/prev  (protected)
func (h *RadioHandlers) SkipPrev(c *gin.Context) {
	if err := h.svc.SkipPrev(); err != nil {
		apiresponse.ErrorFromErr(c, err)
		return
	}
	apiresponse.OK(c, gin.H{})
}

File: web/src/lib/api/radio.ts
import { request, buildQuery } from "./client";

// ---------------------------------------------------------------------------
// Public status
// ---------------------------------------------------------------------------

export interface TrackInfo {
    id: number;
    title: string;
    artist: string;
    album: string;
    genre: string;
    year?: number;
    trackNum?: number;
    format?: string;
    duration: number;
    checksum: string;
    filePath: string;
    coverUrl?: string;
}

export interface RadioStatus {
    station_name: string;
    current_track: string;
    current_track_info: TrackInfo | null;
    total_tracks: number;
    active_clients: number;
    max_clients: number;
    active_tag: string;
    active_playlist: string;
    active_playlist_id: number | null;
    scheduler_running: boolean;
    playlist_summary: Record<string, number>;
    library_tracks: number;
    timezone: string;
    server_time: string;
}

export async function getStatus(channel?: string): Promise<RadioStatus> {
    return request<RadioStatus>("GET", `/api/status${buildQuery({ channel })}`, null, { noAuth: true });
}

export async function getHealth(): Promise<unknown> {
    return request("GET", "/health", null, { noAuth: true });
}

// ---------------------------------------------------------------------------
// Queue
// ---------------------------------------------------------------------------

export async function getQueue(channel?: string): Promise<{ tracks: TrackInfo[] }> {
    return request<{ tracks: TrackInfo[] }>("GET", `/api/queue${buildQuery({ channel })}`, null, { noAuth: true });
}

// ---------------------------------------------------------------------------
// Skip controls (protected)
// ---------------------------------------------------------------------------

export async function skipNext(channel?: string): Promise<void> {
    await request("POST", `/api/skip/next${buildQuery({ channel })}`);
}

export async function skipPrev(channel?: string): Promise<void> {
    await request("POST", `/api/skip/prev${buildQuery({ channel })}`);
}

// ---------------------------------------------------------------------------
// Scheduler
// ---------------------------------------------------------------------------

export interface SchedulerStatus {
    running: boolean;
    last_tag: string;
    current_tag: string;
    time_tags: string[];
    library_tracks?: number;
    summary: Record<string, unknown>;
}

export async function getSchedulerStatus(channel?: string): Promise<SchedulerStatus> {
    return request<SchedulerStatus>("GET", `/api/scheduler/status${buildQuery({ channel })}`, null, { noAuth: true });
}

// ---------------------------------------------------------------------------
// Timezone
// ---------------------------------------------------------------------------

export async function getTimezone(): Promise<{ timezone: string }> {
    return request<{ timezone: string }>("GET", "/api/timezone", null, { noAuth: true });
}

export async function setTimezone(timezone: string): Promise<{ timezone: string }> {
    return request<{ timezone: string }>("PUT", "/api/timezone", { timezone });
}

// ---------------------------------------------------------------------------
// Reconcile / hot-reload
// ---------------------------------------------------------------------------

export interface ReconcileResult {
    removed_count: number;
    orphaned_count: number;
    total_tracks?: number;
}

export async function reconcile(): Promise<ReconcileResult> {
    return request<ReconcileResult>("POST", "/api/reconcile");
}
